lib/ts/jemallctl.cc

```diff
@@ -0,0 +1,336 @@
+/** @file
+
+  Memory allocation routines for libts
+
+  @section license License
+
+  Licensed to the Apache Software Foundation (ASF) under one
+  or more contributor license agreements.  See the NOTICE file
+  distributed with this work for additional information
+  regarding copyright ownership.  The ASF licenses this file
+  to you under the Apache License, Version 2.0 (the
+  "License"); you may not use this file except in compliance
+  with the License.  You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+ */
+
+#include "ts/ink_platform.h"
+#include "ts/ink_memory.h"
+
+// uses config flags
+#include "ts/jemallctl.h"
+
+// includes jemalloc.h if present..
+#include "ts/ink_memory.h"
+#include "ts/ink_defs.h"
+#include "ts/ink_stack_trace.h"
+#include "ts/Diags.h"
+#include "ts/ink_atomic.h"
+#include "ts/ink_align.h"
+
+#include <cassert>
+#if defined(linux) && !defined(_XOPEN_SOURCE)
+#define _XOPEN_SOURCE 600
+#endif
+
+#include <vector>
+#include <cstdlib>
+#include <cstring>
+
+#include <string>
+
+#include <sys/syscall.h> /* For SYS_xxx definitions */
+
+namespace jemallctl
+{
+// instantiations here
+template struct GetObjFxn<uint64_t>;
+template struct GetObjFxn<uint64_t *>;
+template struct GetObjFxn<std::atomic_ulong *>;
+template struct GetObjFxn<unsigned>;
+template struct GetObjFxn<bool>;
+template struct GetObjFxn<bool *>;
+template struct GetObjFxn<std::string>;
+
+template struct SetObjFxn<uint64_t>;
+template struct SetObjFxn<unsigned>;
+template struct SetObjFxn<bool>;
+template struct SetObjFxn<std::string>;
+
+// some inlines need implementing after all
+template uint64_t GetObjFxn<uint64_t>::operator()() const;
+template uint64_t *GetObjFxn<uint64_t *>::operator()() const;
+template std::atomic_ulong *GetObjFxn<std::atomic_ulong *>::operator()() const;
+template unsigned GetObjFxn<unsigned>::operator()() const;
+template bool GetObjFxn<bool>::operator()() const;
+template bool *GetObjFxn<bool *>::operator()() const;
+template std::string GetObjFxn<std::string>::operator()() const;
```

Comment:
I'm getting "duplicate explicit instantiation" errors for lines 68~74 when compiling, on both Ubuntu 17 and RHEL 7. Commenting out these lines makes the error goes away. I see you added this quite recently, why are they needed?

Reply:
I believe the version of GCC I used needed that for some reason (4.8.5. on Centos here and 6.4.1 at home here).  I found it strange because I needed those to *explicitly* instantiate those *member* functions ... yet I'd already instantiated their classes above.

If I/you can remove those .. I'd prefer it personally.  I'll try again there.